### ISSUE TITLE
Getting stdin from flake8's pep8 won't work on 3.x

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#516]
In GitLab by @xZise on Jul 7, 2016, 02:51

Using Flake8 3.x the `pep8` (or actually `pycodestyle`) import yields only an empty string. So while plugins can use `flake8.engine.pep8.get_stdin_value()` in Flake8 2.x this won't work anymore in Flake8 3.x.

My assumption is that now Flake8 itself caches the value and so when a plugin asks `pep8` about it, it has been consumed already. See also the [PR discussion in `flake8-quotes`](https://github.com/zheller/flake8-quotes/pull/37).

I guess the basic class would be:

```
from flake8.engine import pep8

class TestPlugin(object):

    def __init__(self, tree, filename):
        print(repr(pep8.stdin_get_value()))
```

As I understand it Flake8 3.x isn't as dependent on pep8 as Flake8 2.x, but writing a plugin supporting both versions becomes more tedious than just account for the parameters (as you can see in the pull request for the changes needed). I'm very much in favor of the new implementation, that I can just the lines in the constructor but that doesn't work in Flake8 2.x.

The result using the class from above:
```
$ echo "a = 42" | flake8 --ignore FI - ; flake8 --version
3.0.0b2 (pyflakes: 1.2.3, pycodestyle: 2.0.0, flake8_quotes: 0.5.0, flake8-future-import: 0.4.3, mccabe: 0.5.0) CPython 3.5.1 on Linux
[…]
$ echo "a = 42" | flake8 --ignore FI - ; flake8 --version
'a = 42\n'
2.6.2 (pycodestyle: 2.0.0, flake8_quotes: 0.5.0, flake8-future-import: 0.4.3, mccabe: 0.5.0, flake8-docstrings: 0.2.8, pep257: 0.7.0, pyflakes: 1.2.3) CPython 3.5.2 on Linux
```

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jul 7, 2016, 09:07

So, I'm not sure why you all ever imported `pep8` from `flake8.engine`. We monkey-patched that function so you would only have to import pep8 after flake8.

If it makes it easier, as much as I hate the idea, we can monkey-patch `pycodestyle`'s function to make writing this code simpler.

I'm also planning to export a `get_stdin_value` function in Flake8 3.0's API for people who think they need it in 3.0 although they really really shouldn't.

**Comment by asottile** [comment]
In GitLab by @xZise on Jul 8, 2016, 06:10

The main reason I reported the bug is, that it is either a regression (as it'll break other people's builds) or it should be documented that using `pep8`/`pycodestyle` is not supported anymore. I'm fine with breaking compatibility but other developers should know what your plan is, especially as my patch is stalled because it might not be needed when this is a genuine bug.

Now as I cannot find the documentation, I'm asking myself: What should I do? I did a few tests based on [0.5.1 of flake8-quotes](https://github.com/zheller/flake8-quotes/tree/0.5.1) which is only importing `pep8` and using `pep8.stdin_get_value()`. Here are my results:

* Flake8 2.4.0 (pep8): returns empty string (no `pycodestyle` test done)
* Flake8 2.6.2 (pycodestyle): `stdin_get_value` raises `ValueError` (traceback below) importing `pep8`
* Flake8 2.6.2 (pycodestyle): works importing `pycodestyle`
* Flake8 3.0.0b2 (pycodestyle): returns empty string

Now you mention “We monkey-patched that function so you would only have to import pep8 after flake8.”, but in none of my tests importing `pep8` worked when I imported `flake8` first. And only of test worked when importing `pycodestyle`, but there you mention that this is not monkey patched. So I'm very confused about what you try to tell me there.

You wondered why I imported `pep8` from `flake8.engine`? Because it'll be `pycodestyle` or `pep8`, whatever is used by Flake8. And that works in more cases than just importing `pep8`. I could import  `pycodestyle` first and `pep8` if the first import failed, but why when Flake8 is already exposing which package it is using.

But the main problem is that I cannot find a documentation on how to get `stdin` using Flake8. And `flake8-quotes` just imported `pep8`, which didn't work and it was attributed to changing to `pycodestyle`. So I tried to guess what works and I assumed that, as Flake8 is now using `pycodestyle`, that `pep8` just couldn't be used.

The `ValueError` looks as follows:
```
$ echo 'this_should_be_linted = "double quote string"' | flake8 - ; flake8 --version
Traceback (most recent call last):
  File "/home/xzise/.pyenv/versions/3.5.2/bin/flake8", line 9, in <module>
    load_entry_point('flake8', 'console_scripts', 'flake8')()
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/flake8/main.py", line 36, in main
    report = flake8_style.check_files()
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/flake8/engine.py", line 181, in check_files
    return self._retry_serial(self._styleguide.check_files, paths=paths)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/flake8/engine.py", line 172, in _retry_serial
    return func(*args, **kwargs)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/pycodestyle.py", line 1877, in check_files
    runner(path)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/flake8/engine.py", line 126, in input_file
    return fchecker.check_all(expected=expected, line_offset=line_offset)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/pycodestyle.py", line 1608, in check_all
    self.check_ast()
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/pycodestyle.py", line 1555, in check_ast
    for lineno, offset, text, check in checker.run():
  File "/home/xzise/Programms/flake8-quotes/flake8_quotes/__init__.py", line 81, in run
    file_contents = self.get_file_contents()
  File "/home/xzise/Programms/flake8-quotes/flake8_quotes/__init__.py", line 76, in get_file_contents
    return pep8.stdin_get_value().splitlines(True)
  File "/home/xzise/.pyenv/versions/3.5.2/lib/python3.5/site-packages/pep8-1.7.0-py3.5.egg/pep8.py", line 1198, in stdin_get_value
    return TextIOWrapper(sys.stdin.buffer, errors='ignore').read()
ValueError: I/O operation on closed file
2.6.2 (pycodestyle: 2.0.0, flake8_quotes: 0.5.1, mccabe: 0.5.0, pyflakes: 1.2.3) CPython 3.5.2 on Linux
```

Sorry for the long post, but this is a very confusing topic for me and I try to understand your comment, but I don't. So I try around to guess what you might mean and so I get a lot of “data”.

**Comment by asottile** [comment]
In GitLab by @remcohaszing on Jul 13, 2016, 01:16

This issue already exists in Flake8 2.6.2.

I figured out this doesn't exist yet in Flake8 2.5.5.

Stacktrace using 2.6.2:

```
$ cat conftest.py | flake8-atom.sh -
  unknown option 'import-order-style' ignored
  unknown option 'application-import-names' ignored
Traceback (most recent call last):
  File "/home/remco/.local/share/virtualenvs/appsemble-api/bin/flake8", line 11, in <module>
    sys.exit(main())
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/flake8/main.py", line 36, in main
    report = flake8_style.check_files()
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/flake8/engine.py", line 181, in check_files
    return self._retry_serial(self._styleguide.check_files, paths=paths)
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/flake8/engine.py", line 172, in _retry_serial
    return func(*args, **kwargs)
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/pycodestyle.py", line 1877, in check_files
    runner(path)
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/flake8/engine.py", line 126, in input_file
    return fchecker.check_all(expected=expected, line_offset=line_offset)
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/pycodestyle.py", line 1608, in check_all
    self.check_ast()
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/pycodestyle.py", line 1555, in check_ast
    for lineno, offset, text, check in checker.run():
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/flake8_quotes/__init__.py", line 80, in run
    file_contents = self.get_file_contents()
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/flake8_quotes/__init__.py", line 75, in get_file_contents
    return pep8.stdin_get_value().splitlines(True)
  File "/home/remco/.local/share/virtualenvs/appsemble-api/lib/python3.5/site-packages/pep8.py", line 1198, in stdin_get_value
    return TextIOWrapper(sys.stdin.buffer, errors='ignore').read()
ValueError: I/O operation on closed file
```

**Comment by asottile** [comment]
In GitLab by @xZise on Jul 13, 2016, 02:58

There are two different issues though.
* I describe here the issue that Flake8 3.x returns nothing (aka empty string) when calling `pycodestyle.stdin_get_value()`, which does work on 2.6 however.
* You describe the issue that when Flake8 uses `pycodestyle` but the plugin uses `pep8` a `ValueError` occurs.

Maybe your issue is work another bug request too, but as I don't know what the official way of reading from `stdin` is, I don't know if I'm doing it wrong or the plugin broke it. But as far as I can see there are four different cases:
* Flake8 <2.5.1 (before the monkey-patch) using pep8
* Flake8 >=2.5<2.6 using pep8
* Flake8 >=2.6 using pycodestyle
* Flake8 3.x using pycodestyle

Each of those need their own handling, and I have working code for each of them apart from the first. To support Flake8 >=2.5.1 I just need to import the same stdin module as Flake8 is importing (e.g. `pep8` or `pycodestyle`). Now the easiest way is to just use `flake8.engine.pep8`, because in theory you could install Flake8 2.5 (with pep8) and then later import `pycodestyle`. And if you just import `pycodestyle` and fall back to `pep8` you probably have the same issue as you have now with just importing `pep8`. To be honest I haven't tested that yet, but if that won't work I'd like to know what to do instead if using `flake8.engine.pep8` is not wanted.

And then to support Flake8 3.x the new custom constructor can be used. But if it is intentional that Flake8 3.x breaks the compatibility, it should be documented, so that plugin developers know that they need to handle that new case.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jul 20, 2016, 06:16

mentioned in merge request !77

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jul 20, 2016, 06:49

Status changed to closed by commit d69cb0564ada817b1cda5af49a552684466eb2dc

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jul 20, 2016, 12:52

mentioned in commit leorochael/flake8@d69cb0564ada817b1cda5af49a552684466eb2dc
